server/src/com/thoughtworks/go/config/update/RoleConfigCommand.java

```diff
@@ -63,73 +57,124 @@ public boolean canContinue(CruiseConfig cruiseConfig) {
         return isAuthorized();
     }
 
-    protected boolean isValidForCreateOrUpdate(CruiseConfig preprocessedConfig) {
-        preprocessedRole = findExistingRole(preprocessedConfig);
-        preprocessedRole.validate(null);
-        validate(preprocessedConfig);
-        BasicCruiseConfig.copyErrors(preprocessedRole, role);
-        final RolesConfig rolesConfig = preprocessedConfig.server().security().getRoles();
-
-        if (preprocessedRole.errors().isEmpty()) {
-            rolesConfig.validate(null);
-            role.errors().addAll(rolesConfig.errors());
-            return preprocessedRole.getAllErrors().isEmpty() && role.errors().isEmpty();
+    @Override
+    public boolean isValid(CruiseConfig preprocessedConfig) {
+        preprocessedRole = preprocessedConfig.server().security().getRoles().findByName(role.getName());
+
+        if (!preprocessedRole.validateTree(validationContextWithSecurityConfig(preprocessedConfig))) {
+            BasicCruiseConfig.copyErrors(preprocessedRole, role);
+            return false;
         }
 
-        return false;
+        return true;
     }
 
-    private void validate(CruiseConfig preprocessedConfig) {
-        if (role instanceof PluginRoleConfig) {
-            PluginRoleConfig role = (PluginRoleConfig) this.role;
-            PluginRoleConfig preprocessedRole = (PluginRoleConfig) this.preprocessedRole;
-
-            SecurityAuthConfig securityAuthConfig = preprocessedConfig.server().security().securityAuthConfigs().find(role.getAuthConfigId());
-            if (securityAuthConfig == null) {
-                role.addError("authConfigId", "No such security auth configuration present " + role.getAuthConfigId());
-                return;
-            }
-
-            try {
-                ValidationResult result = extension.validateRoleConfiguration(securityAuthConfig.getPluginId(), role.getConfigurationAsMap(true));
-                if (!result.isSuccessful()) {
-                    for (ValidationError validationError : result.getErrors()) {
-                        ConfigurationProperty property = preprocessedRole.getProperty(validationError.getKey());
-                        if (property == null) {
-                            role.addNewConfiguration(validationError.getKey(), false);
-                            preprocessedRole.addNewConfiguration(validationError.getKey(), false);
-                            property = preprocessedRole.getProperty(validationError.getKey());
-                        }
-                        property.addError(validationError.getKey(), validationError.getMessage());
-
-                    }
-                }
-            } catch (PluginNotFoundException e) {
-                role.addError("authConfigId", e.getMessage());
+    private ValidationContext validationContextWithSecurityConfig(final CruiseConfig preprocessedConfig) {
+        return new ValidationContext() {
```

Comment:
Is it helpful if we have `DefaultValidationContext` which will have all empty implementation ? 

```java
return new DefaultValidationContext() {
@Override
             public SecurityConfig getServerSecurityConfig() {
                 return preprocessedConfig.server().security();
             }
}
````